File: app/src/main/java/com/hover/stax/ApplicationInstance.java
package com.hover.stax;

import android.app.Application;

import com.appsflyer.AppsFlyerConversionListener;
import com.appsflyer.AppsFlyerLib;
import com.google.firebase.FirebaseApp;
import com.hover.stax.utils.fonts.FontReplacer;
import com.hover.stax.utils.fonts.Replacer;
import com.yariksoffice.lingver.Lingver;

import java.util.Locale;
import java.util.Map;

import timber.log.Timber;

public class ApplicationInstance extends Application {

    @Override
    public void onCreate() {
        super.onCreate();

        setFont();
        FirebaseApp.initializeApp(this);
        setLogger();
        initAppsFlyer();
    }

    private void setFont() {
        Replacer replacer = FontReplacer.Build(getApplicationContext());
        replacer.setBoldFont("Effra_Heavy.ttf");
        replacer.setMediumFont("Effra_Medium.ttf");
        replacer.setDefaultFont("Effra_Regular.ttf");
        replacer.setThinFont("Effra_Regular.ttf");
        replacer.applyFont();

        Lingver.init(this, Locale.getDefault());
    }

    private void setLogger() {
        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree());
        } else {
            Timber.uprootAll();
        }
    }

    private void initAppsFlyer() {
        AppsFlyerConversionListener conversionListener = new AppsFlyerConversionListener() {
            @Override
            public void onConversionDataSuccess(Map<String, Object> conversionData) {
                for (String attrName : conversionData.keySet()) {
                    Timber.d("attribute: " + attrName + " = " + conversionData.get(attrName));
                }
            }
            @Override
            public void onConversionDataFail(String errorMessage) {
                Timber.d("error getting conversion data: %s", errorMessage);
            }
            @Override
            public void onAppOpenAttribution(Map<String, String> attributionData) {
                for (String attrName : attributionData.keySet()) {
                    Timber.d("attribute: " + attrName + " = " + attributionData.get(attrName));
                }
            }
            @Override
            public void onAttributionFailure(String errorMessage) {
                Timber.d("error onAttributionFailure : %s", errorMessage);
            }
        };
        AppsFlyerLib.getInstance().init(getString(R.string.appsflyer_key), conversionListener, this);
    }
}

File: app/src/main/java/com/hover/stax/ApplicationInstance.kt
package com.hover.stax

import android.app.Application
import android.content.ComponentCallbacks
import com.appsflyer.AppsFlyerConversionListener
import com.appsflyer.AppsFlyerLib
import com.appsflyer.AppsFlyerProperties
import com.google.firebase.FirebaseApp
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.hover.sdk.api.Hover
import com.hover.stax.database.appModule
import com.hover.stax.database.dataModule
import com.hover.stax.database.networkModule
import com.hover.stax.utils.network.NetworkMonitor
import com.yariksoffice.lingver.Lingver
import org.koin.android.ext.koin.androidContext
import org.koin.core.context.startKoin
import timber.log.Timber
import java.util.*
import kotlin.properties.Delegates

class ApplicationInstance : Application() {

    override fun onCreate() {
        super.onCreate()

        setLocale()
        initDI()

        setLogger()
        initFirebase()

        initAppsFlyer()
    }

    private fun initFirebase() {
        FirebaseApp.initializeApp(this)
        FirebaseCrashlytics.getInstance().setCrashlyticsCollectionEnabled(!BuildConfig.DEBUG)
    }

    private fun setLocale() {
        Lingver.init(this, Locale.getDefault())
    }

    private fun initDI() {
        startKoin {
            androidContext(this@ApplicationInstance)
            modules(listOf(appModule, dataModule, networkModule))
        }
    }

    private fun setLogger() {
        if (BuildConfig.DEBUG) Timber.plant(Timber.DebugTree()) else Timber.uprootAll()
    }

    private fun initAppsFlyer() {
        val conversionListener = object : AppsFlyerConversionListener {
            override fun onConversionDataSuccess(data: MutableMap<String, Any>?) {
                data?.keys?.forEach {
                    Timber.d("Attribute $it = ${data[it]}")
                }
            }

            override fun onConversionDataFail(errorMessage: String?) = Timber.d("Error getting conversion data: $errorMessage")

            override fun onAppOpenAttribution(data: MutableMap<String, String>?) {
                data?.keys?.forEach {
                    Timber.d("Attribute $it = ${data[it]}")
                }
            }

            override fun onAttributionFailure(errorMessage: String?) = Timber.d("Error onAttributionFailure : $errorMessage")
        }

        AppsFlyerLib.getInstance().apply {
            init(getString(R.string.appsflyer_key), conversionListener, this@ApplicationInstance)

            if(AppsFlyerProperties.getInstance().getString(AppsFlyerProperties.APP_USER_ID) == null)
                setCustomerUserId(Hover.getDeviceId(this@ApplicationInstance))

            start(this@ApplicationInstance)
        }
    }


    companion object {
        val txnDetailsRetryCounter: MutableMap<String, Int> by Delegates.observable(HashMap()) { _, _, _ -> }
    }

    override fun registerComponentCallbacks(callback: ComponentCallbacks?) {
        super.registerComponentCallbacks(callback)
        NetworkMonitor(this).startNetworkCallback()
    }

    override fun unregisterComponentCallbacks(callback: ComponentCallbacks?) {
        super.unregisterComponentCallbacks(callback)
        NetworkMonitor(this).stopNetworkCallback()
    }
}